### PR TITLE
Fix inspector lifecycle on hot restart bug and add a regression test.

### PR DESCRIPTION
**Pull request by jacob314**

### CODE REVIEW

File: packages/devtools/lib/src/inspector/inspector.dart [hunk 1]
```diff
@@ -104,12 +103,6 @@ class InspectorScreen extends Screen {
       // Init the inspector service, or return null.
       inspectorService =
           await InspectorService.create(service).catchError((e) => null);
-      final pubRootDirectory =
```
Comment (jacob314) [inline packages/devtools/lib/src/inspector/inspector.dart]: This code needed to be moved so it would apply each time we hot restarted.

File: packages/devtools/lib/src/inspector/inspector_service.dart [hunk 2]
```diff
@@ -110,10 +87,13 @@ class InspectorService {
           return functionNames;
         }
       }
-      throw Exception('WidgetInspectorService class not found');
+      // WidgetInspectorService is not available. Either this is not a Flutter
```
Comment (jacob314) [inline packages/devtools/lib/src/inspector/inspector_service.dart]: Unrelated fix. Reduce the amount of console spamming in profile mode.

File: packages/devtools/lib/src/inspector/inspector_service.dart [hunk 3]
```diff
@@ -241,7 +221,7 @@ class InspectorService {
   }
 
   void onExtensionVmServiceRecieved(Event e) {
-    if ('Flutter.Frame' == e.kind) {
+    if ('Flutter.Frame' == e.extensionKind) {
```
Comment (jacob314) [inline packages/devtools/lib/src/inspector/inspector_service.dart]: flat out bug. we didn't need to listen for the flutter frame because it happened that the app would always start after the first frame was already rendered so this code path didn't matter.

File: packages/devtools/lib/src/inspector/inspector_tree_canvas.dart [hunk 4]
```diff
@@ -172,6 +172,8 @@ class InspectorTreeNodeCanvasRender
   // TODO(jacobr): share this method with test code.
   @override
   Icon hitTest(Offset location) {
+    if (offset == null) return null;
```
Comment (jacob314) [inline packages/devtools/lib/src/inspector/inspector_tree_canvas.dart]: slightly related edge case that hit while debugging a broken version of this cl.

File: packages/devtools/test/inspector_controller_test.dart [hunk 5]
```diff
@@ -544,5 +544,71 @@ void main() async {
...
+              '      │     ▼[/icons/inspector/textArea.png] Text\n'
+              '      └─▼[A] AppBar\n'
+              '          ▼[/icons/inspector/textArea.png] Text\n',
+        ),
+      );
+
+      await serviceManager.performHotRestart();
+      // The isolate starts out paused on a hot restart so we have to resume
+      // it manually to make the test pass.
+      await serviceManager.service
```
Comment (jacob314) [inline packages/devtools/test/inspector_controller_test.dart]: @devoncarew pauseOnResume after restart is probably an edge case we haven't thought through fully yet for general devtools-debugger integration. I'm not sure if there is a way to implement it currently without things getting quite odd if you have multiple debuggers attached.

File: packages/devtools/lib/src/utils.dart [hunk 6]
```diff
@@ -279,6 +279,7 @@ class RateLimiter {
   void _performRequest() async {
     try {
       _lastRequestTime = DateTime.now().millisecondsSinceEpoch;
+      _pendingRequest = Completer();
```
Comment (jacob314) [inline packages/devtools/lib/src/utils.dart]: This class was broken but we didn't notice due to the bug in reporting flutter.frame events.

File: packages/devtools/test/inspector_controller_test.dart (hunk 5 above)
Comment (devoncarew) [reply packages/devtools/test/inspector_controller_test.dart]: Definitely would require some thought. Either coordination about who's in charge of resuming, or a convention that you only resume if you're the tool that launched the app in the first place.

File: packages/devtools/lib/src/inspector/inspector_service.dart (hunk 3 above)
Comment (devoncarew) [reply packages/devtools/lib/src/inspector/inspector_service.dart]: Are both `kind` and `extensionKind` strings? If no, we should enable the lint that only allows comparisons between same types.

Comment (jacob314) [reply packages/devtools/lib/src/inspector/inspector_service.dart]: Unfortunately they are both strings but it looks like at one point we wanted to make kind be an enum type.
```dart
/*EventKind*/ String kind;
```
We do have `test_types_in_equals` enabled. I'm happy to know this lint exists as that is often something I worry about when writing an `==`

File: packages/devtools/test/inspector_controller_test.dart (hunk 5 above)
Comment (jacob314) [reply packages/devtools/test/inspector_controller_test.dart]: Some other ideas:
1. Eliminate the pause altogether. I'm not clear why we need pause on resume after a hot restart. Shouldn't the VM just maintain the same list of breakpoints that existed before the hot restart? The pause on initial app start case seems like it should only have 1 debugger connected so isn't a problem.
2. Only resume after all VM Service clients that have opted into pause on start behavior have asked to resume. Would require adding an extra vm service method for a client to opt in that it needs to be notified on resumes after hot restart.

File: packages/devtools/lib/src/inspector/inspector_tree_canvas.dart (hunk 4 above)
Comment (kenzieschmoll) [inline packages/devtools/lib/src/inspector/inspector_tree_canvas.dart]: I don't have a preference either way, but it seems we have a mix of inlined returns and returns in brackets throughout the codebase. We should probably pick a style and stay consistent with it.

File: packages/devtools/lib/src/inspector/inspector_controller.dart [hunk 7]
```diff
@@ -482,11 +482,10 @@ class InspectorController implements InspectorServiceClient {
       // Wait for the master to update.
       return;
     }
-    updateSelectionFromService();
+    updateSelectionFromService(false);
```
Comment (kenzieschmoll) [inline packages/devtools/lib/src/inspector/inspector_controller.dart]: Do we not have the unawaited future lint on anymore? Shouldn't it be showing a warning for this line?

File: packages/devtools/test/inspector_controller_test.dart [hunk 8]
```diff
@@ -544,5 +544,75 @@ void main() async {
...
-  }, tags: 'useFlutterSdk', timeout: const Timeout.factor(2));
+
+    test('hotRestart', () async {
+      await env.setupEnvironment();
+
+      // Select row index 4.
+      // The important thing about this is that the details tree should scroll
+      // instead of re-rooting as the selected row is already visible in the
+      // details tree.
+      tree.onTap(const Offset(0, rowHeight * 4.5));
```
Comment (kenzieschmoll) [inline packages/devtools/test/inspector_controller_test.dart]: probably me just not being super familiar with this test, but where does 4.5 come from?

File: packages/devtools/lib/src/inspector/inspector_controller.dart [hunk 9]
```diff
@@ -482,11 +482,10 @@ class InspectorController implements InspectorServiceClient {
...
   }
 
-  Future<void> updateSelectionFromService() async {
-    final bool firstFrame = !treeLoadStarted;
+  Future<void> updateSelectionFromService(bool firstFrame) async {
```
[the remark this answers is not in the document]
Comment (jacob314) [reply packages/devtools/lib/src/inspector/inspector_controller.dart]: done. I agree that is better.

File: packages/devtools/lib/src/inspector/inspector_tree_canvas.dart (hunk 4 above)
Comment (jacob314) [reply packages/devtools/lib/src/inspector/inspector_tree_canvas.dart]: Lets follow the dart style guide which supports this sort of style for returns. We should cleanup other cases to match this style where it makes sense.
In general, only trivial returns should be on their own line and typically they should only occur near the top of methods so the control flow isn't confusing.

File: packages/devtools/lib/src/inspector/inspector_controller.dart (hunk 7 above)
Comment (jacob314) [reply packages/devtools/lib/src/inspector/inspector_controller.dart]: that lint only warns if the method calling the async method also happens to be async. in this case it isn't
btw in InspectorService unawaited futures are actually fine for the most part because the ObjectGroups are used as a hammer used to abandon obsolete work and wait for pending work to complete.

File: packages/devtools/test/inspector_controller_test.dart (hunk 8 above)
Comment (jacob314) [reply packages/devtools/test/inspector_controller_test.dart]: just in the middle of row index 4.

Comment (jacob314) [reply packages/devtools/test/inspector_controller_test.dart]: added a 
tapRow helper to make that self documenting.

Comment (jacob314) [reply packages/devtools/test/inspector_controller_test.dart]: correction. clickRow
